nanocoat/lib/scritchui/scritchPencilRasterShapes.c

```diff
@@ -17,7 +17,399 @@
+
+		/* If style is DOTTED, rendering will paint and skip pixels 1 by 1. */
+		if (dot)
+			error |= drawPixel(g, innerX, innerY);
+		dot ^= dotFlip;
+	}
+
+	/* Failed? */
+	if (sjme_error_is(error))
+		goto fail_any;
+		
+	/* Success? */
+	return error;
+	
+fail_any:
+	
+	return sjme_error_default(error);
+}
+
+sjme_errorCode sjme_scritchpen_corePrim_fillArc(
+	sjme_attrInNotNull sjme_scritchui_pencil g,
+	sjme_attrInValue sjme_jint x,
+	sjme_attrInValue sjme_jint y,
+	sjme_attrInPositive sjme_jint w,
+	sjme_attrInPositive sjme_jint h,
+	sjme_attrInValue sjme_jint startAngle,
+	sjme_attrInValue sjme_jint arcAngle)
+{
+	sjme_errorCode error;
+	sjme_scritchui_pencilDrawHorizFunc drawHoriz;
+	sjme_jint yz, yze, steps, innerX, innerY;
+	sjme_jboolean hasAlpha = 0;
+	float centerX, centerY, radiusX, radiusY, startAngleRad, endAngleRad,
+		maxRad, angle;
+	sjme_jboolean *filledPixels = NULL;
+	sjme_scritchui_line* clipLine;
+	sjme_scritchui_pencilDrawPixelFunc drawPixel;
+
+
+	if (g == NULL)
+		return SJME_ERROR_NULL_ARGUMENTS;
+		
+	error = SJME_ERROR_NONE;
+
+	/* Java's coordinate system has positive angles moving counter-clockwise */
+	arcAngle = -arcAngle;
+	startAngle = -startAngle;
+
+	hasAlpha = g->hasAlpha;
+
+	drawPixel = g->prim.drawPixel;
+
+	/* Get clipping information. */
+	clipLine = &g->state.clipLine;
+
+	/** 
+	 * Only allocate the alpha buffer if the color isn't opaque. Noticeably
+	 * improves performance for opaque arcs.
+	 *
+	 * TODO: This can be optimized to take 1/8 of the usual memory size by
+	 * manipulating each bit of the boolean array below, as a _Bool variable
+	 * is often at least one whole byte in size.
+	 */
+	if(hasAlpha)
+	{
+		filledPixels = (sjme_jboolean *) sjme_alloca((w * h) *
+			sizeof(sjme_jboolean));
+		if (filledPixels == NULL)
+		{
+			error = sjme_error_outOfMemory(NULL, (w * h) *
+				sizeof(sjme_jboolean));
+			goto fail_any;
+		}
+		memset(filledPixels, 0, (w * h) * sizeof(sjme_jboolean));
+	}
+
+	/**
+	 * This is just drawArc's Bresenham midpoint circle algorithm modified to 
+	 * draw arcs from the center to the edge and check if a pixel was already
+	 * painted before. Works great but is slow on larger transparent arcs.
+	 * Trying to use a scanline approach could result in better performance due
+	 * to the simpler checks (just go from yMax to yMin while making sure a y 
+	 * position is never used twice) but doesn't seem to work right in some
+	 * edge cases (negative angles, too wide/narrow arc angles, etc) so this
+	 * one, albeit slow, is what we'll be going with at the moment.
+	 * 
+	 * At least it has the upside of generating far more stable arcs that fully
+	 * match their outline compared to Java AWT's algorithm.
+	 * 
+	 * TODO: Optimize this later
+	 */ 
+	
+	centerX = x + w / 2.0f;
+	centerY = y + h / 2.0f;
+	radiusX = w / 2.0f;
+	radiusY = h / 2.0f;
+	startAngleRad = startAngle * 0.017453292f;
+	endAngleRad = ((startAngle + arcAngle) * 0.017453292f) 
+	- startAngleRad;
```

Comment:
`#define` the magic number `0.017453292f`.

`-` should be on the previous line and `startAngleRad` should be a continuation indent.